File: Burning Tree.cpp
/*
class Node {
  public:
    int data;
    Node *left;
    Node *right;

    Node(int val) {
        data = val;
        left = right = NULL;
    }
};
*/
class Solution {
  public:
    int minTime(Node* root, int target) {
        // parents needed because tree burn upward direction  also
        map<Node*, Node*>parents;
        queue<Node*>q1;
        q1.push(root);
        Node* targetnode = NULL;
        while(!q1.empty())
        {
            int s = q1.size();
            for(int i=0;i<s;i++)
            {
                
                Node* node = q1.front();
                q1.pop();
                
                // get here target node by using target value
                if(node->data == target)targetnode = node;
                
                if(node->left)
                {
                    parents[node->left] = node;
                    q1.push(node->left);
                }
                if(node->right)
                {
                    parents[node->right] = node;
                    q1.push(node->right);
                }
                
            }
            
        }
            
        map<Node*, bool>burned;
        queue<Node*>q2;
        q2.push(targetnode);
        burned[targetnode] = true;
        int required_sec = -1;
            
        while(!q2.empty())
        {
            int s = q2.size();
            required_sec++;
            for(int i=0;i<s;i++)
            {
                Node* current = q2.front();
                q2.pop();
                
                if(!burned[current->left]&&current->left)
                {
                    q2.push(current->left);
                    burned[current->left] = true;
                }
                
                if(!burned[current->right]&&current->right)
                {
                    q2.push(current->right);
                    burned[current->right] = true;
                }
                
                if(!burned[parents[current]]&&parents[current])
                {
                    q2.push(parents[current]);
                    burned[parents[current]] = true;
                }
            }
        }
                
        return required_sec;  
    }
};

File: Burning Tree.java
/*
class Node {
    int data;
    Node left;
    Node right;

    Node(int data) {
        this.data = data;
        left = null;
        right = null;
    }
}
*/

class Solution {
    public int minTime(Node root, int target) {
        // code here
        if(root==null)
        return 0;
        
        Map<Integer,List<Integer>> graph = new HashMap<>();
        buildGraph(root,null,graph);
        
        int time=0;
        Queue<Integer>q= new LinkedList<>();
        Set<Integer>visited= new HashSet<>();
        q.add(target);
        visited.add(target);
        
        while(!q.isEmpty())
        {
            boolean burned=false;
            int size=q.size();
            for(int i=0;i<size;i++)
            {
                
                for(int neighbour:graph.get(q.poll()))
                {
                    if(!visited.contains(neighbour))
                    {
                       visited.add(neighbour);
                       q.add(neighbour);
                       burned=true;
                    }
                }
            }
            if(burned)
            time++;
        }
        
       return  time; 
    }
    
    public void buildGraph(Node root,Node parent,Map<Integer,List<Integer>>graph)
    {
        if(root==null)
        return;
        
        if(!graph.containsKey(root.data))
        {
            graph.put(root.data,new ArrayList<>());
        }
        if(parent!=null)
        {
              graph.get(root.data).add(parent.data);
            graph.get(parent.data).add(root.data);
        }
        
        buildGraph(root.left,root,graph);
        buildGraph(root.right,root,graph);
    }
}

